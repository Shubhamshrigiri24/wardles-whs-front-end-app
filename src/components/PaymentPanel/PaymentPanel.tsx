import React from "react";
import {
  Typography,
  Button,
  Grid,
  Box,
  makeStyles,
  Theme,
  CircularProgress,
  Container,
  Card,
} from "@material-ui/core";
import { Alert } from "@welldigital/components";

const useStyles = makeStyles((theme: Theme) => ({
  loadingSpinnerCenter: {
    display: "flex",
    justifyContent: "center",
    alignItems: "center",
  },
  marginRight: {
    marginRight: theme.spacing(1),
  },
  pageMargin: {
    marginTop: theme.spacing(2),
    marginBottom: theme.spacing(15),
    [theme.breakpoints.down("sm")]: {
      margin: theme.spacing(0),
    },
  },
  card: {
    border: "none",
    background: "none",
    [theme.breakpoints.down("sm")]: {
      margin: theme.spacing(0),
      padding: "0px",
    },
  },
}));

export interface PaymentPanelStateProps {
  paymentErrorMessage: string;
  disableButton: boolean;
  loading: boolean;
  isPaying: boolean;
}

export interface PaymentPanelDispatchProps {
  onPay(): void;
}

export type PaymentPanelProps = PaymentPanelStateProps &
  PaymentPanelDispatchProps;

const braintreeContainerId = "braintree_container";
export const BRAINTREE_CONTAINER_ID = `#${braintreeContainerId}`;

const PaymentPanel: React.FC<PaymentPanelProps> = ({
  disableButton,
  isPaying,
  loading,
  onPay,
  paymentErrorMessage,
}) => {
  const classes = useStyles();
  return (
    <Container maxWidth="sm" className={classes.pageMargin}>
      <Card className={classes.card}>
        <Typography variant="h3">Select a method of payment</Typography>
        <div id={braintreeContainerId} />
        {paymentErrorMessage && (
          <Alert type={"error"} message={paymentErrorMessage} />
        )}
        <Grid container>
          <Box mt={4}>
            <Button
              variant="contained"
              data-testid="paymentPanel/payment-button"
              color="primary"
              disabled={disableButton || isPaying}
              onClick={onPay}
            >
              {isPaying || loading ? (
                <div className={classes.loadingSpinnerCenter}>
                  <CircularProgress
                    size={"1em"}
                    className={classes.marginRight}
                  />
                  Paying...
                </div>
              ) : (
                "Pay"
              )}
            </Button>
          </Box>
        </Grid>
      </Card>
    </Container>
  );
};

export default PaymentPanel;
