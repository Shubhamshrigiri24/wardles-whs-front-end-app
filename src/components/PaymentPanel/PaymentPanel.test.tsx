import React from "react";
import { render, fireEvent, act } from "@testing-library/react";
import "@testing-library/jest-dom/extend-expect";
import PaymentPanel, { PaymentPanelProps } from "./PaymentPanel";

jest.mock("react-router-dom", () => ({
  useHistory: () => ({
    push: mockHistoryPush,
  }),
}));

const mockHistoryPush = jest.fn();
beforeEach(() => {
  mockHistoryPush.mockReset();
});

describe("<PaymentPanel />", () => {
  it("should match the snapshot", async () => {
    const props = makeProps();
    await act(async () => {
      const { container } = render(<PaymentPanel {...props} />);
      expect(container).toMatchSnapshot();
    });
  });
  it("should match the snapshot on error", async () => {
    const props = makeProps({ paymentErrorMessage: "error" });
    await act(async () => {
      const { container } = render(<PaymentPanel {...props} />);
      expect(container).toMatchSnapshot();
    });
  });
  it("should match the snapshot when loading", async () => {
    const props = makeProps({ loading: true });
    await act(async () => {
      const { container } = render(<PaymentPanel {...props} />);
      expect(container).toMatchSnapshot();
    });
  });
  it("should call onPay if you click the pay button", async () => {
    const props = makeProps();
    const { getByTestId } = render(<PaymentPanel {...props} />);
    await act(async () => {
      fireEvent.click(getByTestId("paymentPanel/payment-button"));
    });
    expect(props.onPay).toHaveBeenCalled();
  });
  it("should not call onPay the button is disabled", async () => {
    const props = makeProps({ disableButton: true });
    const { getByTestId } = render(<PaymentPanel {...props} />);
    await act(async () => {
      fireEvent.click(getByTestId("paymentPanel/payment-button"));
    });
    expect(props.onPay).toHaveBeenCalledTimes(0);
  });
});

function makeProps(v?: Partial<PaymentPanelProps>): PaymentPanelProps {
  return {
    paymentErrorMessage: v?.paymentErrorMessage ?? "",
    disableButton: v?.disableButton ?? false,
    loading: v?.loading ?? false,
    onPay: jest.fn(),
    isPaying: false,
  };
}
