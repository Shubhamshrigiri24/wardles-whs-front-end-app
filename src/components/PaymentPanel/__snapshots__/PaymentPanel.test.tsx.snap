// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<PaymentPanel /> should match the snapshot 1`] = `
<div>
  <div
    class="MuiContainer-root makeStyles-pageMargin-3 MuiContainer-maxWidthSm"
  >
    <div
      class="MuiPaper-root MuiCard-root makeStyles-card-4 MuiPaper-elevation1 MuiPaper-rounded"
    >
      <h3
        class="MuiTypography-root MuiTypography-h3"
      >
        Select a method of payment
      </h3>
      <div
        id="braintree_container"
      />
      
      <div
        class="MuiGrid-root MuiGrid-container"
      >
        <div
          class="MuiBox-root MuiBox-root-5"
        >
          <button
            class="MuiButtonBase-root MuiButton-root MuiButton-contained MuiButton-containedPrimary"
            data-testid="paymentPanel/payment-button"
            tabindex="0"
            type="button"
          >
            <span
              class="MuiButton-label"
            >
              Pay
            </span>
          </button>
        </div>
      </div>
    </div>
  </div>
</div>
`;

exports[`<PaymentPanel /> should match the snapshot on error 1`] = `
<div>
  <div
    class="MuiContainer-root makeStyles-pageMargin-8 MuiContainer-maxWidthSm"
  >
    <div
      class="MuiPaper-root MuiCard-root makeStyles-card-9 MuiPaper-elevation1 MuiPaper-rounded"
    >
      <h3
        class="MuiTypography-root MuiTypography-h3"
      >
        Select a method of payment
      </h3>
      <div
        id="braintree_container"
      />
      <div
        class="makeStyles-root-10 makeStyles-errorType-14 makeStyles-defaultSize-16"
        data-testid="alert/root"
      >
        <div
          class="makeStyles-main-20"
          data-testid="alert/main"
        >
          <div
            class="makeStyles-icon-21"
            data-testid="alert/icon"
          >
            <svg
              aria-hidden="true"
              class="MuiSvgIcon-root"
              focusable="false"
              viewBox="0 0 24 24"
            >
              <path
                d="M12 2C6.48 2 2 6.48 2 12s4.48 10 10 10 10-4.48 10-10S17.52 2 12 2zm1 15h-2v-2h2v2zm0-4h-2V7h2v6z"
              />
            </svg>
          </div>
          <p
            class="MuiTypography-root makeStyles-message-23 MuiTypography-body1"
            data-testid="alert/message"
          >
            error
          </p>
        </div>
      </div>
      <div
        class="MuiGrid-root MuiGrid-container"
      >
        <div
          class="MuiBox-root MuiBox-root-27"
        >
          <button
            class="MuiButtonBase-root MuiButton-root MuiButton-contained MuiButton-containedPrimary"
            data-testid="paymentPanel/payment-button"
            tabindex="0"
            type="button"
          >
            <span
              class="MuiButton-label"
            >
              Pay
            </span>
          </button>
        </div>
      </div>
    </div>
  </div>
</div>
`;

exports[`<PaymentPanel /> should match the snapshot when loading 1`] = `
<div>
  <div
    class="MuiContainer-root makeStyles-pageMargin-30 MuiContainer-maxWidthSm"
  >
    <div
      class="MuiPaper-root MuiCard-root makeStyles-card-31 MuiPaper-elevation1 MuiPaper-rounded"
    >
      <h3
        class="MuiTypography-root MuiTypography-h3"
      >
        Select a method of payment
      </h3>
      <div
        id="braintree_container"
      />
      
      <div
        class="MuiGrid-root MuiGrid-container"
      >
        <div
          class="MuiBox-root MuiBox-root-32"
        >
          <button
            class="MuiButtonBase-root MuiButton-root MuiButton-contained MuiButton-containedPrimary"
            data-testid="paymentPanel/payment-button"
            tabindex="0"
            type="button"
          >
            <span
              class="MuiButton-label"
            >
              <div
                class="makeStyles-loadingSpinnerCenter-28"
              >
                <div
                  class="MuiCircularProgress-root makeStyles-marginRight-29 MuiCircularProgress-colorPrimary MuiCircularProgress-indeterminate"
                  role="progressbar"
                  style="width: 1em; height: 1em;"
                >
                  <svg
                    class="MuiCircularProgress-svg"
                    viewBox="22 22 44 44"
                  >
                    <circle
                      class="MuiCircularProgress-circle MuiCircularProgress-circleIndeterminate"
                      cx="44"
                      cy="44"
                      fill="none"
                      r="20.2"
                      stroke-width="3.6"
                    />
                  </svg>
                </div>
                Paying...
              </div>
            </span>
          </button>
        </div>
      </div>
    </div>
  </div>
</div>
`;
